### ISSUE TITLE
Output filenames containing the string ".bcf" result in BCF output instead of VCF

### ISSUE THREAD

**Issue by jrandall** [samtools/htsjdk#761]
### Subject of the issue
Output filenames containing the string ".bcf" anywhere in the filename result in BCF output when VCF output is desired. For example, if one has filenames containing an md5 offset by dots, and an md5 happens to start with "bcf" then you can wind up with unexpected output format.

### Expected behaviour

An output filename such as `library.bcfce04b6b0fba5bd286ed9a1c9ebfe1.g.vcf.gz` should produce VCF output.

### Actual behaviour

An output filename such as `library.bcfce04b6b0fba5bd286ed9a1c9ebfe1.g.vcf.gz` produces BCF output.

**Comment by jrandall** [comment]
Relevant code is here: https://github.com/samtools/htsjdk/blob/fbba5364e1809de071bc479f30e4e2c8b17f5bbe/src/main/java/htsjdk/variant/variantcontext/writer/VariantContextWriterFactory.java#L252

Possibly the solution could be changing `.contains()` to `.endsWith()` but I'm not sure why that wasn't done in the first place. Are there some BCF files that are expected to have a suffix beyond `.bcf` ?

**Comment by nh13** [comment]
Likely it was there to support `.bcf.gz`, but the variant package is a bit mysterious to me.

**Comment by magicDGS** [comment]
That class is deprecated, and actually I did a PR to remove it (https://github.com/samtools/htsjdk/pull/756). Does this happen when using `VariantContextWriterBuilder`, @jrandall?

**Comment by magicDGS** [comment]
Checking the no-deprecated class, this should not happen (see this: https://github.com/samtools/htsjdk/blob/fbba5364e1809de071bc479f30e4e2c8b17f5bbe/src/main/java/htsjdk/variant/variantcontext/writer/VariantContextWriterBuilder.java#L498).

**Comment by jrandall** [comment]
@magicDGS I don't use htsjdk as a developer - I ran into this problem using the GATK and tracked it down to this class. It looks like the GATK is still using the deprecated class: https://github.com/broadgsa/gatk-protected/blob/f185a75e1c49fb4c039511e61254da0509833ee9/public/gatk-engine/src/main/java/org/broadinstitute/gatk/engine/io/storage/VariantContextWriterStorage.java#L139

**Comment by nh13** [comment]
@vdauwera given this ultimately shows up in GATK, you may want to take a look.

**Comment by vdauwera** [comment]
@ronlevine Could you please check if we're still using this in the dev snapshot? If it's an easy change I'd love to get this fixed for the 3.7 release, even if it means delaying a day or two.

**Comment by ronlevine** [comment]
It's in the dev snapshot.  So, the logic should be changed from:
`location.getName().contains(".bcf")`
to
 `endsWith(".bcf.gz") || endsWith(".bcf")`.

**Comment by vdauwera** [comment]
@ronlevine To be clear I think we would want to update to using the newer class mentioned above -- especially given @magicDGS 's PR to remove the old class entirely -- though perhaps the new class needs to be updated to allow bcf.gz.

**Comment by ronlevine** [comment]
@vdauwera The logic snippet pertains to the non-deprecated class https://github.com/samtools/htsjdk/blob/fbba5364e1809de071bc479f30e4e2c8b17f5bbe/src/main/java/htsjdk/variant/variantcontext/writer/VariantContextWriterBuilder.java#L498

**Comment by vdauwera** [comment]
Oh ok

**Comment by ronlevine** [comment]
In GATK's [VariantContextWriterStorage](https://github.com/broadgsa/gatk-protected/blob/f185a75e1c49fb4c039511e61254da0509833ee9/public/gatk-engine/src/main/java/org/broadinstitute/gatk/engine/io/storage/VariantContextWriterStorage.java#L139), replace `VariantContextWriterFactory` with the appropriate`VariantContextWriterBuilder` logic

**Comment by ronlevine** [comment]
On second thought, there should be no need to make a change in HTSJDK since any file ending in `.gz` is handled as [compressed VCF](https://github.com/samtools/htsjdk/blob/master/src/main/java/htsjdk/variant/variantcontext/writer/VariantContextWriterBuilder.java#L474)

**Comment by vdauwera** [comment]
> On second thought, there should be no need to make a change in HTSJDK since any file ending in .gz, .gzip, .bgz, .bgzf is handled as compressed VCF

Is that actually desired behavior? Doesn't that mean we can't write a bcf.gz, unless there is logic within the "compressed VCF" handling to allow the BCF case?

**Comment by ronlevine** [comment]
No, it isn't, `bcf.gz` is written as `vcf.gz`. I created a test on the branch `rhl_validate_bcf.gz` to investigate this behavior;

**Comment by ronlevine** [comment]
@yfarjoun Do you have any comments on this one? If not, I think this should be closed.

**Comment by yfarjoun** [comment]
have we resolved the issue? I think that the issue is not an HTSJDK issue, but rather a GATK issue that was solved...if that's the case, I'm happy for this issue to be closed...but let's ask the OP: @jrandall ?

**Comment by tfenne** [comment]
Fixed on master: the deprecated VariantContextWriterFactory that used .contains(".bcf") was removed (#756), and VariantContextWriterBuilder now detects BCF via endsWith(".bcf") (VariantContextWriterBuilder#isBCF). A filename like library.bcf....g.vcf.gz now produces VCF output. Closing.
